xds/src/main/java/io/grpc/xds/LoadStatsManager2.java

```diff
@@ -0,0 +1,396 @@
+import io.grpc.xds.EnvoyProtoData.ClusterStats.DroppedRequests;
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.EnvoyProtoData.UpstreamLocalityStats;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Manages client side traffic stats. Drop stats are maintained in cluster (with edsServiceName)
+ * granularity and load stats (request counts) are maintained in locality granularity.
+ */
+@ThreadSafe
+final class LoadStatsManager2 {
+  // Recorders for drops of each cluster:edsServiceName.
+  private final Map<String, Map<String, ReferenceCounted<ClusterDropStats>>> allDropStats =
+      new HashMap<>();
+  // Recorders for loads of each cluster:edsServiceName:locality.
+  private final Map<String, Map<String,
+      Map<Locality, ReferenceCounted<ClusterLocalityStats>>>> allLoadStats = new HashMap<>();
+  private final Supplier<Stopwatch> stopwatchSupplier;
+
+  LoadStatsManager2(Supplier<Stopwatch> stopwatchSupplier) {
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+  }
+
+  /**
+   * Gets or creates the stats object for recording drops for the specified cluster with
+   * edsServiceName. The returned object is reference counted and the caller should use {@link
+   * ClusterDropStats#release()} to release its <i>hard</i> reference when it is safe to discard
+   * future stats for the cluster.
+   */
+  synchronized ClusterDropStats getClusterDropStats(
+      String cluster, @Nullable String edsServiceName) {
+    if (!allDropStats.containsKey(cluster)) {
+      allDropStats.put(cluster, new HashMap<String, ReferenceCounted<ClusterDropStats>>());
+    }
+    Map<String, ReferenceCounted<ClusterDropStats>> perClusterCounters = allDropStats.get(cluster);
+    if (!perClusterCounters.containsKey(edsServiceName)) {
+      perClusterCounters.put(
+          edsServiceName,
+          ReferenceCounted.wrap(new ClusterDropStats(
+              cluster, edsServiceName, stopwatchSupplier.get())));
+    }
+    ReferenceCounted<ClusterDropStats> ref = perClusterCounters.get(edsServiceName);
+    ref.retain();
+    return ref.get();
+  }
+
+  private synchronized void releaseClusterDropCounter(
+      String cluster, @Nullable String edsServiceName) {
+    checkState(allDropStats.containsKey(cluster)
+            && allDropStats.get(cluster).containsKey(edsServiceName),
+        "stats for cluster %s, edsServiceName %s not exits", cluster, edsServiceName);
+    ReferenceCounted<ClusterDropStats> ref = allDropStats.get(cluster).get(edsServiceName);
+    ref.release();
```

Comment:
Calling stats object's `release()` does not immediately deletes the counter. It just decrements the ref count. The counter will be discarded after next load report. See line 162.